component/src/stake/component.rs

```diff
@@ -549,18 +562,16 @@ impl Staking {
 
     /// Returns the list of validator updates formatted for inclusion in the Tendermint `EndBlockResponse`
     pub async fn tm_validator_updates(&self) -> Result<Vec<ValidatorUpdate>> {
-        // Tracking validator updates by identity key rather than by consensus
-        // key is convenient internally, but to create the Tendermint update, we
-        // now need to look up the consensus key for each validator.
         let mut updates = Vec::new();
-        for (identity_key, power) in &self.tm_validator_updates {
-            let validator = self
-                .state
-                .validator(identity_key)
-                .await?
-                .ok_or_else(|| anyhow::anyhow!("queued update for missing validator"))?;
+        for (consensus_key, power) in &self.tm_validator_updates {
+            if self.new_consensus_keys.contains(consensus_key) && *power == 0 {
+                // tendermint can't handle receiving 0 voting power for a new
+                // consensus key
+                continue;
+            }
+
             updates.push(ValidatorUpdate {
-                pub_key: validator.consensus_key,
+                pub_key: *consensus_key,
                 power: (*power).try_into().unwrap(),
             });
         }
```

Comment:
We can avoid the allocation here I think, by instead using a `filter_map` on an iterator over the updates, returning an `impl Iterator`.